src/machine/machine_nrf52840.go

```diff
@@ -33,3 +35,72 @@ var (
 	PWM2 = &PWM{PWM: nrf.PWM2}
 	PWM3 = &PWM{PWM: nrf.PWM3}
 )
+
+// PDM represents a PDM device
+type PDM struct {
+	device        *nrf.PDM_Type
+	defaultBuffer int16
+}
+
+// Configure is intended to set up the PDM interface prior to use.
+func (pdm *PDM) Configure(config PDMConfig) error {
+	if config.DIN == 0 {
+		return errors.New("No DIN pin provided in configuration")
+	}
+
+	if config.CLK == 0 {
+		return errors.New("No CLK pin provided in configuration")
+	}
+
+	config.DIN.Configure(PinConfig{Mode: PinInput})
+	config.CLK.Configure(PinConfig{Mode: PinOutput})
+	pdm.device = nrf.PDM
+	pdm.device.PSEL.DIN.Set(uint32(config.DIN))
+	pdm.device.PSEL.CLK.Set(uint32(config.CLK))
+	pdm.device.PDMCLKCTRL.Set(nrf.PDM_PDMCLKCTRL_FREQ_Default)
+	pdm.device.RATIO.Set(nrf.PDM_RATIO_RATIO_Ratio64)
+	pdm.device.GAINL.Set(nrf.PDM_GAINL_GAINL_DefaultGain)
+	pdm.device.GAINR.Set(nrf.PDM_GAINR_GAINR_DefaultGain)
+	pdm.device.ENABLE.Set(nrf.PDM_ENABLE_ENABLE_Enabled)
+
+	if config.Stereo {
+		pdm.device.MODE.Set(nrf.PDM_MODE_OPERATION_Stereo | nrf.PDM_MODE_EDGE_LeftRising)
+	} else {
+		pdm.device.MODE.Set(nrf.PDM_MODE_OPERATION_Mono | nrf.PDM_MODE_EDGE_LeftRising)
+	}
+
+	pdm.device.SAMPLE.SetPTR(uint32(uintptr(unsafe.Pointer(&pdm.defaultBuffer))))
+	pdm.device.SAMPLE.SetMAXCNT_BUFFSIZE(1)
+	pdm.device.SetTASKS_START(1)
+	return nil
+}
+
+// Read stores a set of samples in the given target buffer. Pointer should
+// represent first element of buffer slice, not the pointer to the slice itself.
+func (pdm *PDM) Read(bufptr *int16, bufsz uint32) (uint32, error) {
```

Comment:
Perhaps you can change the signature to:

```go
func (pdm *PDM) Read(buf []int16) (uint32, error)
```

And then perform the pointer casting inside the function itself. What do you think?

Reply:
I'm not a golang expert, I just hack :) My impression was that passing the value would result in more memory consumed, and the audio buffers are usually large. If that's not the case I'll try it, I agree it is cleaner.

Now that I think about it you're just passing the slice reference itself and not the backing array, so I guess it does make sense. Stand by...